src/legacy/core_plugins/embeddable_api/public/actions/action.ts

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { EuiContextMenuItemIcon } from '@elastic/eui';
+
+import { IEmbeddable } from '../embeddables';
+
+export interface ActionContext<
+  TEmbeddable extends IEmbeddable = IEmbeddable,
+  TTriggerContext extends {} = {}
+> {
+  embeddable: TEmbeddable;
+  triggerContext?: TTriggerContext;
+}
+
+export abstract class Action<
+  TEmbeddable extends IEmbeddable = IEmbeddable,
+  TTriggerContext extends {} = {}
+> {
+  /**
+   * Determined the order when there is more than one action matched to a trigger.
+   * Higher numbers are displayed first.
+   */
+  public order: number = 0;
+  public abstract readonly type: string;
+
+  constructor(public readonly id: string) {}
+
+  /**
+   * Optional icon that can be displayed along with the title.
+   */
+  public getIcon(
```

Comment:
> The icon \ display text \ href etc are used before the action is triggered (i.e. in the context menu \ hover \ etc). So should it depend on the context?

No they are used when the action is triggered.  I think you missed my Embeddable API session at GAH, but if you look at the gif on this slide: https://docs.google.com/presentation/d/1IEWgktt9aUgtqnOUvcEeQRMo3ZFcB1ezE1Rm5FBiUeY/edit#slide=id.g514a35735d_0_202

you can see that the context menu pops up after the user did something that emitted the trigger, if it is tied to more than one action.

<img width="864" alt="Screen Shot 2019-06-03 at 11 06 07 AM" src="https://user-images.githubusercontent.com/16563603/58812355-b1500480-85ef-11e9-815a-ec26f103bbe6.png">

The CONTEXT_MENU_TRIGGER just has no dynamic context, so that trigger is different.